[@types/lodash]: pickBy has incorrect signature when using it with Dictionary

- [x] I tried using the `@types/lodash` package and had problems. (version 4.14.104)
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).

After upgrading to 4.14.104, the error appears when trying to filter a `Dictionary<number>` with `_.pickBy` and reassign to the same variable:
```
error TS2322: Type 'Partial<Dictionary<number>>' is not assignable to type 'Dictionary<number>'.
  Index signatures are incompatible.
    Type 'number | undefined' is not assignable to type 'number'.
      Type 'undefined' is not assignable to type 'number'.
```
The code is similar to this:
```typescript
let mapping: Dictionary<number> = {key1: 5, key2: -3}
mapping = _.pickBy(mapping, v => v > 0)
```
Ideally, the result of `_.pickBy` has to be also a `Dictionary<T>` in this case. 
After all, the new object is created which doesn't have some of the keys.
It is not the same as the object that has some of the values changed to `undefined`.

@andy-ms @bczengel @chrootsu @stepancar @aj-r @Ailrun @e-cloud @thorn0 @jtmthf @DomiR

--- a/types/lodash/common/object.d.ts
+++ b/types/lodash/common/object.d.ts
@@ -2951,6 +2951,22 @@ declare module "../index" {
          * _.omitBy(object, _.isNumber);
          * // => { 'b': '2' }
          */
+        omitBy<T>(
+            object: Dictionary<T> | null | undefined,
+            predicate?: ValueKeyIteratee<T>
+        ): Dictionary<T>;
+
+        /**
+         * @see _.omitBy
+         */
+        omitBy<T>(
+            object: NumericDictionary<T> | null | undefined,
+            predicate?: ValueKeyIteratee<T>
+        ): NumericDictionary<T>;
+
+        /**
+         * @see _.omitBy
+         */
         omitBy<T extends object>(
             object: T | null | undefined,
             predicate: ValueKeyIteratee<T[keyof T]>
@@ -2961,6 +2977,22 @@ declare module "../index" {
         /**
          * @see _.omitBy
          */
+        omitBy<T>(
+            this: LoDashImplicitWrapper<Dictionary<T> | null | undefined>,
+            predicate?: ValueKeyIteratee<T>
+        ): LoDashImplicitWrapper<Dictionary<T>>;
+
+        /**
+         * @see _.omitBy
+         */
+        omitBy<T>(
+            this: LoDashImplicitWrapper<NumericDictionary<T> | null | undefined>,
+            predicate?: ValueKeyIteratee<T>
+        ): LoDashImplicitWrapper<NumericDictionary<T>>;
+
+        /**
+         * @see _.omitBy
+         */
         omitBy<T extends object>(
             this: LoDashImplicitWrapper<T | null | undefined>,
             predicate: ValueKeyIteratee<T[keyof T]>
@@ -2968,6 +3000,22 @@ declare module "../index" {
     }
 
     interface LoDashExplicitWrapper<TValue> {
+        /**
+         * @see _.omitBy
+         */
+        omitBy<T>(
+            this: LoDashExplicitWrapper<Dictionary<T> | null | undefined>,
+            predicate?: ValueKeyIteratee<T>
+        ): LoDashExplicitWrapper<Dictionary<T>>;
+
+        /**
+         * @see _.omitBy
+         */
+        omitBy<T>(
+            this: LoDashExplicitWrapper<NumericDictionary<T> | null | undefined>,
+            predicate?: ValueKeyIteratee<T>
+        ): LoDashExplicitWrapper<NumericDictionary<T>>;
+
         /**
          * @see _.omitBy
          */
@@ -3063,6 +3111,22 @@ declare module "../index" {
          * _.pickBy(object, _.isNumber);
          * // => { 'a': 1, 'c': 3 }
          */
+        pickBy<T>(
+            object: Dictionary<T> | null | undefined,
+            predicate?: ValueKeyIteratee<T>
+        ): Dictionary<T>;
+
+        /**
+         * @see _.pickBy
+         */
+        pickBy<T>(
+            object: NumericDictionary<T> | null | undefined,
+            predicate?: ValueKeyIteratee<T>
+        ): NumericDictionary<T>;
+
+        /**
+         * @see _.pickBy
+         */
         pickBy<T extends object>(
             object: T | null | undefined,
             predicate?: ValueKeyIteratee<T[keyof T]>
@@ -3073,6 +3137,22 @@ declare module "../index" {
         /**
          * @see _.pickBy
          */
+        pickBy<T>(
+            this: LoDashImplicitWrapper<Dictionary<T> | null | undefined>,
+            predicate?: ValueKeyIteratee<T>
+        ): LoDashImplicitWrapper<Dictionary<T>>;
+
+        /**
+         * @see _.pickBy
+         */
+        pickBy<T>(
+            this: LoDashImplicitWrapper<NumericDictionary<T> | null | undefined>,
+            predicate?: ValueKeyIteratee<T>
+        ): LoDashImplicitWrapper<NumericDictionary<T>>;
+
+        /**
+         * @see _.pickBy
+         */
         pickBy<T extends object>(
             this: LoDashImplicitWrapper<T | null | undefined>,
             predicate?: ValueKeyIteratee<T[keyof T]>
@@ -3080,6 +3160,22 @@ declare module "../index" {
     }
 
     interface LoDashExplicitWrapper<TValue> {
+        /**
+         * @see _.pickBy
+         */
+        pickBy<T>(
+            this: LoDashExplicitWrapper<Dictionary<T> | null | undefined>,
+            predicate?: ValueKeyIteratee<T>
+        ): LoDashExplicitWrapper<Dictionary<T>>;
+
+        /**
+         * @see _.pickBy
+         */
+        pickBy<T>(
+            this: LoDashExplicitWrapper<NumericDictionary<T> | null | undefined>,
+            predicate?: ValueKeyIteratee<T>
+        ): LoDashExplicitWrapper<NumericDictionary<T>>;
+
         /**
          * @see _.pickBy
          */

--- a/types/lodash/fp.d.ts
+++ b/types/lodash/fp.d.ts
@@ -2289,11 +2289,21 @@ declare namespace _ {
     }
     interface LodashOmitBy {
         <T>(predicate: lodash.ValueKeyIteratee<T>): LodashOmitBy1x1<T>;
-        <T extends object>(predicate: lodash.__, object: T | null | undefined): LodashOmitBy1x2<T>;
+        <T>(predicate: lodash.__, object: lodash.Dictionary<T> | null | undefined): LodashOmitBy1x2<T>;
+        <T>(predicate: lodash.ValueKeyIteratee<T>, object: lodash.Dictionary<T> | null | undefined): lodash.Dictionary<T>;
+        <T>(predicate: lodash.__, object: lodash.NumericDictionary<T> | null | undefined): LodashOmitBy2x2<T>;
+        <T>(predicate: lodash.ValueKeyIteratee<T>, object: lodash.NumericDictionary<T> | null | undefined): lodash.NumericDictionary<T>;
+        <T extends object>(predicate: lodash.__, object: T | null | undefined): LodashOmitBy3x2<T>;
         <T extends object>(predicate: lodash.ValueKeyIteratee<T[keyof T]>, object: T | null | undefined): lodash.PartialObject<T>;
     }
-    type LodashOmitBy1x1<T> = <T1 extends object>(object: T1 | null | undefined) => lodash.PartialObject<T1>;
-    type LodashOmitBy1x2<T> = (predicate: lodash.ValueKeyIteratee<T[keyof T]>) => lodash.PartialObject<T>;
+    interface LodashOmitBy1x1<T> {
+        (object: lodash.Dictionary<T> | null | undefined): lodash.Dictionary<T>;
+        (object: lodash.NumericDictionary<T> | null | undefined): lodash.NumericDictionary<T>;
+        <T1 extends object>(object: T1 | null | undefined): lodash.PartialObject<T1>;
+    }
+    type LodashOmitBy1x2<T> = (predicate: lodash.ValueKeyIteratee<T>) => lodash.Dictionary<T>;
+    type LodashOmitBy2x2<T> = (predicate: lodash.ValueKeyIteratee<T>) => lodash.NumericDictionary<T>;
+    type LodashOmitBy3x2<T> = (predicate: lodash.ValueKeyIteratee<T[keyof T]>) => lodash.PartialObject<T>;
     type LodashOnce = <T extends (...args: any[]) => any>(func: T) => T;
     interface LodashOrderBy {
         <T>(iteratees: lodash.Many<(value: T) => lodash.NotVoid>): LodashOrderBy1x1<T>;
@@ -2873,11 +2883,21 @@ declare namespace _ {
     type LodashPick2x2<T> = (props: lodash.PropertyPath) => lodash.PartialDeep<T>;
     interface LodashPickBy {
         <T>(predicate: lodash.ValueKeyIteratee<T>): LodashPickBy1x1<T>;
-        <T extends object>(predicate: lodash.__, object: T | null | undefined): LodashPickBy1x2<T>;
+        <T>(predicate: lodash.__, object: lodash.Dictionary<T> | null | undefined): LodashPickBy1x2<T>;
+        <T>(predicate: lodash.ValueKeyIteratee<T>, object: lodash.Dictionary<T> | null | undefined): lodash.Dictionary<T>;
+        <T>(predicate: lodash.__, object: lodash.NumericDictionary<T> | null | undefined): LodashPickBy2x2<T>;
+        <T>(predicate: lodash.ValueKeyIteratee<T>, object: lodash.NumericDictionary<T> | null | undefined): lodash.NumericDictionary<T>;
+        <T extends object>(predicate: lodash.__, object: T | null | undefined): LodashPickBy3x2<T>;
         <T extends object>(predicate: lodash.ValueKeyIteratee<T[keyof T]>, object: T | null | undefined): lodash.PartialObject<T>;
     }
-    type LodashPickBy1x1<T> = <T1 extends object>(object: T1 | null | undefined) => lodash.PartialObject<T1>;
-    type LodashPickBy1x2<T> = (predicate: lodash.ValueKeyIteratee<T[keyof T]>) => lodash.PartialObject<T>;
+    interface LodashPickBy1x1<T> {
+        (object: lodash.Dictionary<T> | null | undefined): lodash.Dictionary<T>;
+        (object: lodash.NumericDictionary<T> | null | undefined): lodash.NumericDictionary<T>;
+        <T1 extends object>(object: T1 | null | undefined): lodash.PartialObject<T1>;
+    }
+    type LodashPickBy1x2<T> = (predicate: lodash.ValueKeyIteratee<T>) => lodash.Dictionary<T>;
+    type LodashPickBy2x2<T> = (predicate: lodash.ValueKeyIteratee<T>) => lodash.NumericDictionary<T>;
+    type LodashPickBy3x2<T> = (predicate: lodash.ValueKeyIteratee<T[keyof T]>) => lodash.PartialObject<T>;
     interface LodashProp {
         <TObject extends object, TKey extends keyof TObject>(path: TKey | [TKey]): LodashProp1x1<TObject, TKey>;
         <TObject extends object>(path: lodash.__, object: TObject): LodashProp1x2<TObject>;

--- a/types/lodash/lodash-tests.ts
+++ b/types/lodash/lodash-tests.ts
@@ -5561,13 +5561,23 @@ fp.now(); // $ExpectType number
 // _.omitBy
 {
     const obj: AbcObject | null | undefined = anything;
+    const dictionary: _.Dictionary<boolean> | null | undefined = anything;
+    const numericDictionary: _.NumericDictionary<boolean> | null | undefined = anything;
     const predicate = (element: string | number | boolean, key: string) => true;
+    const predicate2 = (element: boolean, key: string) => true;
 
     _.omitBy(obj, predicate); // $ExpectType Partial<AbcObject>
+    _.omitBy(dictionary, predicate2); // $ExpectType Dictionary<boolean>
+    _.omitBy(numericDictionary, predicate2); // $ExpectType NumericDictionary<boolean>
     _(obj).omitBy(predicate); // $ExpectType LoDashImplicitWrapper<Partial<AbcObject>>
+    _(dictionary).omitBy(predicate2); // $ExpectType LoDashImplicitWrapper<Dictionary<boolean>>
+    _(numericDictionary).omitBy(predicate2); // $ExpectType LoDashImplicitWrapper<NumericDictionary<boolean>>
     _.chain(obj).omitBy(predicate); // $ExpectType LoDashExplicitWrapper<Partial<AbcObject>>
+    _.chain(dictionary).omitBy(predicate2); // $ExpectType LoDashExplicitWrapper<Dictionary<boolean>>
+    _.chain(numericDictionary).omitBy(predicate2); // $ExpectType LoDashExplicitWrapper<NumericDictionary<boolean>>
     fp.omitBy(predicate, obj); // $ExpectType Partial<AbcObject>
-    fp.omitBy(predicate)(obj); // $ExpectType Partial<AbcObject>
+    fp.omitBy(predicate2)(dictionary); // $ExpectType Dictionary<boolean>
+    fp.omitBy(predicate2)(numericDictionary); // $ExpectType NumericDictionary<boolean>
 }
 
 // _.pick
@@ -5584,8 +5594,10 @@ fp.now(); // $ExpectType number
     _.pick(obj1, readonlyArray); // $ExpectType PartialDeep<AbcObject>
     // Broken in TS 2.4
     // _.pick(obj2, "a", "b"); // Pick<AbcObject, "a" | "b">
-    _.pick(obj2, literalsArray); // $ExpectType Pick<AbcObject, "a" | "b">
-    _.pick(obj2, roLiteralsArray); // $ExpectType Pick<AbcObject, "a" | "b">
+    // We can't use ExpectType here because typescript keeps changing what order the types appear.
+    let result1: Pick<AbcObject, "a" | "b">;
+    result1 = _.pick(obj2, literalsArray);
+    result1 = _.pick(obj2, roLiteralsArray);
 
     _(obj1).pick("a"); // $ExpectType LoDashImplicitWrapper<Partial<AbcObject>>
     _(obj1).pick(0, "a"); // $ExpectType LoDashImplicitWrapper<Partial<AbcObject>>
@@ -5593,8 +5605,9 @@ fp.now(); // $ExpectType number
     _(obj1).pick(readonlyArray); // $ExpectType LoDashImplicitWrapper<Partial<AbcObject>>
     // Broken in TS 2.4
     // _(obj2).pick("a", "b"); // LoDashImplicitWrapper<Pick<AbcObject, "a" | "b">>
-    _(obj2).pick(literalsArray); // $ExpectType LoDashImplicitWrapper<Pick<AbcObject, "a" | "b">>
-    _(obj2).pick(roLiteralsArray); // $ExpectType LoDashImplicitWrapper<Pick<AbcObject, "a" | "b">>
+    let result2: _.LoDashImplicitWrapper<Pick<AbcObject, "a" | "b">>;
+    result2 = _(obj2).pick(literalsArray);
+    result2 = _(obj2).pick(roLiteralsArray);
 
     _.chain(obj1).pick("a"); // $ExpectType LoDashExplicitWrapper<Partial<AbcObject>>
     _.chain(obj1).pick(0, "a"); // $ExpectType LoDashExplicitWrapper<Partial<AbcObject>>
@@ -5602,24 +5615,35 @@ fp.now(); // $ExpectType number
     _.chain(obj1).pick(readonlyArray); // $ExpectType LoDashExplicitWrapper<Partial<AbcObject>>
     // Broken in TS 2.4
     // _.chain(obj2).pick("a", "b"); // LoDashExplicitWrapper<Pick<AbcObject, "a" | "b">>
-    _.chain(obj2).pick(literalsArray); // $ExpectType LoDashExplicitWrapper<Pick<AbcObject, "a" | "b">>
-    _.chain(obj2).pick(roLiteralsArray); // $ExpectType LoDashExplicitWrapper<Pick<AbcObject, "a" | "b">>
+    let result3: _.LoDashExplicitWrapper<Pick<AbcObject, "a" | "b">>;
+    result3 = _.chain(obj2).pick(literalsArray);
+    result3 = _.chain(obj2).pick(roLiteralsArray);
 
     fp.pick<AbcObject, "a">("a", obj2); // $ExpectType Pick<AbcObject, "a">
     fp.pick<AbcObject, "a">("a")(obj2); // $ExpectType Pick<AbcObject, "a">
-    fp.pick<AbcObject, "a" | "b">(literalsArray)(obj2); // $ExpectType Pick<AbcObject, "a" | "b">
+    result1 = fp.pick<AbcObject, "a" | "b">(literalsArray)(obj2);
 }
 
 // _.pickBy
 {
     const obj: AbcObject | null | undefined = anything;
+    const dictionary: _.Dictionary<boolean> | null | undefined = anything;
+    const numericDictionary: _.NumericDictionary<boolean> | null | undefined = anything;
     const predicate = (element: string | number | boolean, key: string) => true;
+    const predicate2 = (element: boolean, key: string) => true;
 
     _.pickBy(obj, predicate); // $ExpectType Partial<AbcObject>
+    _.pickBy(dictionary, predicate2); // $ExpectType Dictionary<boolean>
+    _.pickBy(numericDictionary, predicate2); // $ExpectType NumericDictionary<boolean>
     _(obj).pickBy(predicate); // $ExpectType LoDashImplicitWrapper<Partial<AbcObject>>
+    _(dictionary).pickBy(predicate2); // $ExpectType LoDashImplicitWrapper<Dictionary<boolean>>
+    _(numericDictionary).pickBy(predicate2); // $ExpectType LoDashImplicitWrapper<NumericDictionary<boolean>>
     _.chain(obj).pickBy(predicate); // $ExpectType LoDashExplicitWrapper<Partial<AbcObject>>
+    _.chain(dictionary).pickBy(predicate2); // $ExpectType LoDashExplicitWrapper<Dictionary<boolean>>
+    _.chain(numericDictionary).pickBy(predicate2); // $ExpectType LoDashExplicitWrapper<NumericDictionary<boolean>>
     fp.pickBy(predicate, obj); // $ExpectType Partial<AbcObject>
-    fp.pickBy(predicate)(obj); // $ExpectType Partial<AbcObject>
+    fp.pickBy(predicate2)(dictionary); // $ExpectType Dictionary<boolean>
+    fp.pickBy(predicate2)(numericDictionary); // $ExpectType NumericDictionary<boolean>
 }
 
 // _.result

--- a/types/lowdb/_lodash.d.ts
+++ b/types/lowdb/_lodash.d.ts
@@ -1273,6 +1273,14 @@ declare module "./index" {
             this: LoDashExplicitSyncWrapper<T | null | undefined>,
             ...paths: Array<_.Many<_.PropertyName>>
         ): LoDashExplicitSyncWrapper<_.PartialObject<T>>;
+        omitBy<T>(
+            this: LoDashExplicitSyncWrapper<_.Dictionary<T> | null | undefined>,
+            predicate?: _.ValueKeyIteratee<T>
+        ): LoDashExplicitSyncWrapper<_.Dictionary<T>>;
+        omitBy<T>(
+            this: LoDashExplicitSyncWrapper<_.NumericDictionary<T> | null | undefined>,
+            predicate?: _.ValueKeyIteratee<T>
+        ): LoDashExplicitSyncWrapper<_.NumericDictionary<T>>;
         omitBy<T extends object>(
             this: LoDashExplicitSyncWrapper<T | null | undefined>,
             predicate: _.ValueKeyIteratee<T[keyof T]>
@@ -1285,6 +1293,14 @@ declare module "./index" {
             this: LoDashExplicitSyncWrapper<T | null | undefined>,
             ...props: _.PropertyPath[]
         ): LoDashExplicitSyncWrapper<_.PartialObject<T>>;
+        pickBy<T>(
+            this: LoDashExplicitSyncWrapper<_.Dictionary<T> | null | undefined>,
+            predicate?: _.ValueKeyIteratee<T>
+        ): LoDashExplicitSyncWrapper<_.Dictionary<T>>;
+        pickBy<T>(
+            this: LoDashExplicitSyncWrapper<_.NumericDictionary<T> | null | undefined>,
+            predicate?: _.ValueKeyIteratee<T>
+        ): LoDashExplicitSyncWrapper<_.NumericDictionary<T>>;
         pickBy<T extends object>(
             this: LoDashExplicitSyncWrapper<T | null | undefined>,
             predicate?: _.ValueKeyIteratee<T[keyof T]>
@@ -2841,6 +2857,14 @@ declare module "./index" {
             this: LoDashExplicitAsyncWrapper<T | null | undefined>,
             ...paths: Array<_.Many<_.PropertyName>>
         ): LoDashExplicitAsyncWrapper<_.PartialObject<T>>;
+        omitBy<T>(
+            this: LoDashExplicitAsyncWrapper<_.Dictionary<T> | null | undefined>,
+            predicate?: _.ValueKeyIteratee<T>
+        ): LoDashExplicitAsyncWrapper<_.Dictionary<T>>;
+        omitBy<T>(
+            this: LoDashExplicitAsyncWrapper<_.NumericDictionary<T> | null | undefined>,
+            predicate?: _.ValueKeyIteratee<T>
+        ): LoDashExplicitAsyncWrapper<_.NumericDictionary<T>>;
         omitBy<T extends object>(
             this: LoDashExplicitAsyncWrapper<T | null | undefined>,
             predicate: _.ValueKeyIteratee<T[keyof T]>
@@ -2853,6 +2877,14 @@ declare module "./index" {
             this: LoDashExplicitAsyncWrapper<T | null | undefined>,
             ...props: _.PropertyPath[]
         ): LoDashExplicitAsyncWrapper<_.PartialObject<T>>;
+        pickBy<T>(
+            this: LoDashExplicitAsyncWrapper<_.Dictionary<T> | null | undefined>,
+            predicate?: _.ValueKeyIteratee<T>
+        ): LoDashExplicitAsyncWrapper<_.Dictionary<T>>;
+        pickBy<T>(
+            this: LoDashExplicitAsyncWrapper<_.NumericDictionary<T> | null | undefined>,
+            predicate?: _.ValueKeyIteratee<T>
+        ): LoDashExplicitAsyncWrapper<_.NumericDictionary<T>>;
         pickBy<T extends object>(
             this: LoDashExplicitAsyncWrapper<T | null | undefined>,
             predicate?: _.ValueKeyIteratee<T[keyof T]>

